sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,258 @@
+        X = np.zeros((len(data), len(feature_dict)))
+        for idx, row in enumerate(data):
+            for feature_name, feature_value in row.items():
+                X[idx, feature_dict[feature_name]] = feature_value
+        return X
+
+    def compute_tf_idf(self, X):
+        """Transform a term frequency matrix into a TF-IDF matrix.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : ndarray of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        import numpy as np
+
+        n_samples, n_features = X.shape
+
+        df = np.count_nonzero(X, axis=0) + 1
+        n_samples += 1
+        idf = np.log(n_samples / df) + 1
+        X_tfidf = X * idf
+        X_tfidf = (X_tfidf.T / np.linalg.norm(X_tfidf, axis=1)).T
+
+        return X_tfidf
+
+    def cosine_similarity(self, X, Y=None):
+        """
+        Compute the cosine similarity between two vectors X and Y.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : ndarray of shape (n_samples_Y, n_features), default=None
+            Input data. If `None`, the output will be the pairwise
+            similarities between all samples in `X`.
+
+        Returns
+        -------
+        cosine_similarity : ndarray of shape (n_samples_X, n_samples_Y)
+            Cosine similarity matrix.
+        """
+        import numpy as np
+
+        if Y is X or Y is None:
+            Y = X
+
+        X_normalized = X / np.linalg.norm(X)
+        if X is Y:
+            Y_normalized = X_normalized
+        else:
+            Y_normalized = Y / np.linalg.norm(Y)
+        similarity = X_normalized @ Y_normalized.T
+
+        return similarity
+
+    def fit(self, file_names):
+        """
+        Compute the similarity matrix of a group of documents.
+
+        Parameters
+        ----------
+        file_names : list or generator of file names.
+
+        Returns
+        -------
+        self : object
+            Fitted recommender.
+        """
+        import numpy as np
+
+        n_examples = self.n_examples
+        min_df = self.min_df
+        if not isinstance(n_examples, numbers.Integral):
+            raise ValueError("n_examples must be an integer")
+        elif n_examples < 1:
+            raise ValueError("n_examples must be strictly positive")
+        if not isinstance(min_df, numbers.Integral):
+            raise ValueError("min_df must be an integer")
+        elif min_df < 1:
+            raise ValueError("min_df must be strictly positive")
+
+        freq_func = self.token_freqs
+        counts_matrix = self.dict_vectorizer(
+            [freq_func(Path(fname).read_text()) for fname in file_names]
+        )
+        doc_appearances = np.sum(counts_matrix, axis=0)
+        mask = doc_appearances >= min_df
```

Comment:
We could further improve memory usage and recommender speed by trimming with a `max_df=0.9` to remove the most frequent words that appear in more than 90% of the examples: the importance of those words/tokens is automatically diminished by the IDF weighting but trimming them before computing the cosine similarity matrix might:

- improve the recommendations even further (a bit);
- more importantly spare some computation resources that would be wasted on very common, and therefore non-informative tokens.

This would amount to some kind of automated stop word list.